appserver/ejb/ejb-http-remoting/client/src/main/java/fish/payara/ejb/http/client/adapter/CompositeClientAdapter.java

```diff
@@ -0,0 +1,207 @@
+ *
+ *    Contributor(s):
+ *    If you wish your version of this file to be governed by only the CDDL or
+ *    only the GPL Version 2, indicate your decision by adding "[Contributor]
+ *    elects to include this software in this distribution under the [CDDL or GPL
+ *    Version 2] license."  If you don't indicate a single choice of license, a
+ *    recipient has the option to distribute your version of this file under
+ *    either the CDDL, the GPL Version 2 or to extend the choice of license to
+ *    its licensees as provided above.  However, if you add GPL Version 2 code
+ *    and therefore, elected the GPL Version 2 license, then the option applies
+ *    only if the new code is made subject to such option by the copyright
+ *    holder.
+ */
+
+package fish.payara.ejb.http.client.adapter;
+
+import javax.naming.Context;
+import javax.naming.NamingException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Supplier;
+
+/**
+ * Adapter that allows composing multiple adapter implementations. {@Ccode CompositeClientAdapter}
+ * invokes all registered adapters in order in which they were registered, and returns first non-empty match for a name.
+ * <p>Downstream adapters are registered by constructing a new {@link Builder}, or method {@link #newBuilder()}</p>
+ * <p>Downstream adapters can also be customized by decorators provided by {@link ClientAdapterCustomizer}, which
+ * allows separating name matching from proxy construction in adapters' implementations.</p>
+ * @see ClientAdapterCustomizer
+ */
+public final class CompositeClientAdapter implements ClientAdapter {
+    private final List<Supplier<? extends ClientAdapter>> adapterSuppliers;
+
+    private CompositeClientAdapter(Builder builder) {
+        this.adapterSuppliers = new ArrayList<>(builder.adapterSuppliers);
+    }
+
+    /**
+     * Find a client adapter for handling specific {@code jndiName}. Composite client adapter will attempt invoking
+     * {@link ClientAdapter#makeLocalProxy(String, Context)} for all registered adapters in order they were
+     * registered in. The iteration ends when
+     * <ol>
+     *     <li>A registered adapter returns non-empty value; or</li>
+     *     <li>A registered adapter throws {@code NamingException}</li>
+     *     <li>All adapters are tried and return empty value</li>
+     * </ol>
+     *
+     * @param jndiName JNDI name that is being looked up
+     * @param remoteContext EJB HTTP Client naming context for delegating the calls
+     * @return adapter object if any of registered client adapters returns one, {@code Optional.empty()} otherwise
+     * @throws NamingException if any of attempted client adapters throws one
+     */
+    @Override
+    public Optional<Object> makeLocalProxy(String jndiName, Context remoteContext) throws NamingException {
+        Optional<ResolutionResult> resolutionResult = adapterSuppliers.stream()
+                .map(Supplier::get)
```

Comment:
Now came to think if it makes much sense to call `Supplier#get` for each invocation of the method. This might create multiple adapters where only one was expected or meant to be created.

Reply:
I realized using suppliers is too meta in this context. `ClientAdapter` is already a factory, so we shouldn't need factory factory for it.